ledger-service/http-json/src/main/scala/com/digitalasset/http/Endpoints.scala

```diff
@@ -70,44 +71,62 @@ class Endpoints(
         (implicit lc => httpResponse(exercise(req)))
       case req @ HttpRequest(POST, Uri.Path("/v1/create-and-exercise"), _, _, _) =>
         (implicit lc => httpResponse(createAndExercise(req)))
+    }: DispatchFun) andThen (f =>
+      lc => Timed.future(metrics.daml.HttpJsonApi.commandSubmissionTimer, f(lc))
+    )
+    val queryDispatch: DispatchFun = ({
       case req @ HttpRequest(POST, Uri.Path("/v1/fetch"), _, _, _) =>
         (implicit lc => httpResponse(fetch(req)))
       case req @ HttpRequest(GET, Uri.Path("/v1/query"), _, _, _) =>
         (implicit lc => httpResponse(retrieveAll(req)))
       case req @ HttpRequest(POST, Uri.Path("/v1/query"), _, _, _) =>
         (implicit lc => httpResponse(query(req)))
+    }: DispatchFun) andThen (f => lc => Timed.future(metrics.daml.HttpJsonApi.queryTimer, f(lc)))
+    val partyManagementDispatch: DispatchFun = ({
       case req @ HttpRequest(GET, Uri.Path("/v1/parties"), _, _, _) =>
         (implicit lc => httpResponse(allParties(req)))
       case req @ HttpRequest(POST, Uri.Path("/v1/parties"), _, _, _) =>
         (implicit lc => httpResponse(parties(req)))
       case req @ HttpRequest(POST, Uri.Path("/v1/parties/allocate"), _, _, _) =>
         (implicit lc => httpResponse(allocateParty(req)))
+    }: DispatchFun) andThen (f =>
+      lc => Timed.future(metrics.daml.HttpJsonApi.partyManagementTimer, f(lc))
+    )
+    val packageManagementDispatch: DispatchFun = ({
       case req @ HttpRequest(GET, Uri.Path("/v1/packages"), _, _, _) =>
         (implicit lc => httpResponse(listPackages(req)))
       // format: off
       case req @ HttpRequest(GET,
-          Uri(_, _, Slash(Segment("v1", Slash(Segment("packages", Slash(Segment(packageId, Empty)))))), _, _),
-          _, _, _) => (implicit lc => downloadPackage(req, packageId))
+        Uri(_, _, Slash(Segment("v1", Slash(Segment("packages", Slash(Segment(packageId, Empty)))))), _, _),
+        _, _, _) => (implicit lc => downloadPackage(req, packageId))
       // format: on
       case req @ HttpRequest(POST, Uri.Path("/v1/packages"), _, _, _) =>
         (implicit lc => httpResponse(uploadDarFile(req)))
+    }: DispatchFun) andThen (f =>
+      lc => Timed.future(metrics.daml.HttpJsonApi.packageManagementTimer, f(lc))
+    )
+    val liveOrHealthDispatch: DispatchFun = {
       case HttpRequest(GET, Uri.Path("/livez"), _, _, _) =>
         _ => Future.successful(HttpResponse(status = StatusCodes.OK))
       case HttpRequest(GET, Uri.Path("/readyz"), _, _, _) =>
         _ => healthService.ready().map(_.toHttpResponse)
     }
     import scalaz.std.partialFunction._, scalaz.syntax.arrow._
-    (dispatch &&& { case r => r }) andThen { case (lcFhr, req) =>
+    ((commandDispatch orElse
+      queryDispatch orElse
+      partyManagementDispatch orElse
+      packageManagementDispatch orElse
+      liveOrHealthDispatch) &&& { case r => r }) andThen { case (lcFhr, req) =>
```

Comment:
If `&&&` and `orElse` do the same thing we should probably choose one and stick to that.

Reply:
Oh didn't know that, many thanks. I personally would prefer `orElse` as this reads better than `&&&`. The latter is in F# btw bitwise-and :sweat_smile:

Reply:
I do prefer `orElse` as well. @S11001001 what is the advantage of `&&&` over `orElse`? I would guess something related to type inference or variance but I really don't know.

Reply:
Just noticed that messed up, I do know the `&&&` operator and it doesn't mean `orElse` in this context. It means `combine`.

Reply:
What is `combine`?

Reply:
This is a "the type is the documentation" situation for sure 🙂 https://hackage.haskell.org/package/base-4.15.0.0/docs/Control-Arrow.html#v:-38--38--38-